### ISSUE TITLE
Build Windows binaries on opensuse for julia

### ISSUE THREAD

**Issue by davidanthoff** [WizardMac/ReadStat#62]
I'm trying to set up a build of windows binaries on build.opensuse.org. Once that is up and running we can integrated that via the WinRPM package into the julia package and have a super simple julia story for this on Windows.

I'm following the advice [here](https://github.com/JuliaLang/WinRPM.jl/issues/28#issuecomment-121844977).

My opensuse package is [here](https://build.opensuse.org/package/show/home:davidanthoff/mingw64-readstat).

But it doesn't work, if I understand the build log properly something is going wrong at the unzip stage already (maybe it doesn't even find the unzip command?). I'm sure this is just some very simple small stupid error, but I can't find it... @tkelman, any chance you could help me out here and take a brief look at my .spec file?

**Comment by tkelman** [comment]
add unzip as a BuildRequires, or use the .tar.gz source file

**Comment by davidanthoff** [comment]
Thanks, @tkelman, that solved that problem!

Now I'm stuck [here](https://build.opensuse.org/package/live_build_log/home:davidanthoff/mingw64-readstat/openSUSE_13.1/x86_64) that the compiler can't find iconv.h. I think that means libiconv needs to somehow be installed. I tried to just add ``libiconv`` as another BuildRequires, but that doesn't work, and it is not clear to me what to try next...

**Comment by tkelman** [comment]
[mingw64-win_iconv](https://build.opensuse.org/package/show/windows:mingw:win64/mingw64-win_iconv)

**Comment by davidanthoff** [comment]
And again, thanks, @tkelman, that did the trick!

The build now succeeds, but when I click "Download package" [here](https://build.opensuse.org/package/show/home:davidanthoff/mingw64-readstat) there is nothing there...

The Rpmlint results seem to complain about a noarch tag, but I don't understand what that means.

**Comment by tkelman** [comment]
might be a consequence of the other warning about spec file capitalization not matching?

**Comment by davidanthoff** [comment]
Alright, I'm at the point where the download rpm file has the ``readstat.exe`` file in it, so something is building an packaging.

Really I need the libreadstat-0.dll, though... I don't seem to be able to grab it, all I get is a "File not found" error.

I think there might be two avenues forward: a) why is libreadstat-0.dll not in the same folder where readstat.exe ends up? This might be an issue with the current makefiles? b) or maybe we can just figure out where it ends up, and then grab it from there...

I also don't understand why the file is actually called libreadstat-0.dll instead of libreadstat.dll...

@evanmiller maybe you might be able to help out here as well?

I just *really* don't understand the whole unix build environment, which makes this tricky ;)

The config for all of this is [here](https://build.opensuse.org/package/show/home:davidanthoff/mingw64-readstat). The latest log for a failed build [here](https://build.opensuse.org/package/live_build_log/home:davidanthoff/mingw64-readstat/openSUSE_13.2/x86_64).

**Comment by tkelman** [comment]
libtool is getting mad about libiconv for some reason...
```
[   54s] /bin/sh ./libtool  --tag=CC   --mode=link x86_64-w64-mingw32-gcc -Wall -O2 -g -pipe -Wall -fexceptions --param=ssp-buffer-size=4 -mms-bitfields -no-undefined -Wl,--exclude-libs=libintl.a -Wl,--exclude-libs=libiconv.a -Wl,--no-keep-memory -o libreadstat.la -rpath /usr/x86_64-w64-mingw32/sys-root/mingw/lib src/libreadstat_la-CKHashTable.lo src/libreadstat_la-readstat_bits.lo src/libreadstat_la-readstat_convert.lo src/libreadstat_la-readstat_dta.lo src/libreadstat_la-readstat_dta_read.lo src/libreadstat_la-readstat_dta_write.lo src/libreadstat_la-readstat_error.lo src/libreadstat_la-readstat_io_unistd.lo src/libreadstat_la-readstat_parser.lo src/libreadstat_la-readstat_por.lo src/libreadstat_la-readstat_por_parse.lo src/libreadstat_la-readstat_rdata.lo src/libreadstat_la-readstat_sas.lo src/libreadstat_la-readstat_sas_catalog.lo src/libreadstat_la-readstat_sas_data.lo src/libreadstat_la-readstat_sav.lo src/libreadstat_la-readstat_sav_parse.lo src/libreadstat_la-readstat_sav_read.lo src/libreadstat_la-readstat_sav_write.lo src/libreadstat_la-readstat_spss.lo src/libreadstat_la-readstat_spss_parse.lo src/libreadstat_la-readstat_value.lo src/libreadstat_la-readstat_variable.lo src/libreadstat_la-readstat_writer.lo -llzma -lz -liconv 
[   54s] 
[   54s] *** Warning: linker path does not have real file for library -liconv.
[   54s] *** I have the capability to make that library automatically link in when
[   54s] *** you link to this library.  But I can only do this if you have a
[   54s] *** shared version of the library, which you do not appear to have
[   54s] *** because I did check the linker path looking for a file starting
[   54s] *** with libiconv and none of the candidates passed a file format test
[   54s] *** using a file magic. Last file checked: /usr/x86_64-w64-mingw32/sys-root/mingw/lib/libiconv.a
[   54s] *** The inter-library dependencies that have been dropped here will be
[   54s] *** automatically added whenever a program is linked with this library
[   54s] *** or is declared to -dlopen it.
[   54s] 
[   54s] *** Since this library must not contain undefined symbols,
[   54s] *** because either the platform does not support them or
[   54s] *** it was explicitly requested with -no-undefined,
[   54s] *** libtool will only create a static version of it.
```

I tend to recommend people start with cmake rather than autotools these days due to stuff like this. Though here it might be a bad interaction with the `-Wl,--exclude-libs=libiconv.a` flag, and I'm not sure who is adding that.

**Comment by davidanthoff** [comment]
I added ``mingw64-win_iconv-dll`` as a BuildRequires, but that doesn't seem to help, even though its spec file seems to indicate that it includes libiconv.dll and iconv.dll...

I'm lost what to try next... @evanmiller would have to comment about cmake vs autotools, or maybe he has some further ideas?

**Comment by tkelman** [comment]
Part of the problem might be that the win_iconv package is deleting its `.dll.a` import library, which I think would help libtool figure things out if it were present. Maybe try branching (opensuse build service equivalent to a github fork) the win_iconv package to your personal project, removing the deletion of the `.dll.a` file and add it to the `-devel` package?

**Comment by davidanthoff** [comment]
How do I make sure my spec file picks up my my branched version of win_iconv and not the original?

**Comment by tkelman** [comment]
I think it should do that automatically, preferring dependencies within the same repository/project over external links.

**Comment by davidanthoff** [comment]
Hm, that doesn't seem to work, I still get the same message...

**Comment by tkelman** [comment]
It probably needs to be in the same https://build.opensuse.org/project/show/home:davidanthoff project

**Comment by davidanthoff** [comment]
I'm slowly making progress. I think the build is now more or less functional, now I'm trying to add this to the build.jl file in [WizardMac/DataRead.jl](https://github.com/WizardMac/DataRead.jl). I'm trying
````julia
 push!(WinRPM.sources, "http://download.opensuse.org/repositories/home:/davidanthoff:/mingw64-readstat/openSUSE_13.2")
````
which I think is the URL that should get my stuff from https://build.opensuse.org/package/show/home:davidanthoff/mingw64-readstat, but that download doesn't work... @tkelman  any pointer what I'm doing wrong here?

**Comment by tkelman** [comment]
How'd you end up resolving the iconv / dll issue?

You probably need a `WinRPM.update()` after modifying `WinRPM.sources`, are you doing that?

**Comment by davidanthoff** [comment]
I added ``home:davidanthoff:branches:windows:mingw:win64/openSUSE_13.2`` as an additional package repository with higher precedence than ``windows:mingw:win64/openSUSE_13.2`` to my openSUSE_13.2 etc. repositories. That was enough to get the build finish. And the binaries in the RPM file that I can download run on my system, as long as I throw the remaining dlls into the same folder (but I assume WinRPM will take care of that).

There is still a problem with iconv, the download from the original repo there has two versions of the dll: libiconv.dll and iconv.dll. The spec file indicates that one is just a renamed copy of the other, but when I open both in dependency walker on my system, it says only one is a valid dll at all... Not sure what is going on there, but I thought that is something I can figure out later.

I would also hope that I can merge my change to the iconv spec file upstream and just get rid of that whole thing...

I am calling ``WinRPM.update()``, when I do that it is complaining that it can't download anything from the url that I specified.

**Comment by tkelman** [comment]
> I would also hope that I can merge my change to the iconv spec file upstream and just get rid of that whole thing...

Yeah hopefully they'd be okay with it if all you changed is adding the .dll.a back in, you can submit a pull-request-equivalent. The not-valid DLL is probably a symbolic link file that won't work for use on Windows. As long as the julia package has both listed as aliases, I think it should try dlopen'ing them and only succeed on the real one.

> it is complaining that it can't download anything from the url that I specified

I think you need to adjust the url then. You've looked at the CoinOptServices example? In that case I had a separate project under my home folder, I think you need to take `mingw64-readstat` out of the url since that's a package name, not part of the project/repository name.

**Comment by davidanthoff** [comment]
Just for my own memory: this is now blocked on two iconv issues:
- They should [include the a.dll files](https://build.opensuse.org/request/show/402827)
- The iconv.dll shipped in the parent repo is [corrupted](https://bugzilla.opensuse.org/show_bug.cgi?id=985282)

Other things that still need to be done:
- Enable 32 bit
- Ideally move this out of my personal opensuse repo to somewhere else. I don't understand the whole orga thing over there, so not clear whether this should be in something mingw, or something in @evanmiller's name, or maybe some julia related account? @tkelman what would you recommend?

**Comment by tkelman** [comment]
We could propose it for inclusion in the official mingw projects so you don't need the modification of `WinRPM.sources` in `deps/build.jl`, but I don't know if this is a notable enough library to be commonly packaged or depended upon. In the upstream project we'd be exposed a little bit more to breaking changes by the maintainers there, but on the flipside, also maybe insulated from accidentally breaking it on our own?

**Comment by davidanthoff** [comment]
> The not-valid DLL is probably a symbolic link file that won't work for use on Windows. As long as the julia package has both listed as aliases, I think it should try dlopen'ing them and only succeed on the real one.

That won't work because the libreadstat.dll actually depends on the iconv.dll, so we'll never open the iconv.dll directly from julia. Lets just hope they fix this upstream.

**Comment by davidanthoff** [comment]
The [response here](https://build.opensuse.org/request/show/402827) suggests that we statically link libiconv instead of using the dlls, at least for the Windows build on OpenSuse. I have no idea how to do that. @tkelman or @evanmiller could you help me out? Probably really simple, right?

**Comment by tkelman** [comment]
I think the easiest way to go about that is replace the `-liconv` when the readstat library gets linked with the full path to the `.a` file. This won't be very general so is probably better to do in your .spec file rather than in the autotools files here. It might be doable by setting `EXTRA_LIBS` when you configure or make.

**Comment by davidanthoff** [comment]
Ok, I tried to just replace ``-liconv`` with ``libiconv.a`` in the configure.ac file for the ``EXTRA_LIBS``. I now get a complaint during the make phase:
````
$ make
  CC       src/libreadstat_la-CKHashTable.lo
  CC       src/libreadstat_la-readstat_bits.lo
  CC       src/libreadstat_la-readstat_convert.lo
  CC       src/libreadstat_la-readstat_dta.lo
  CC       src/libreadstat_la-readstat_dta_read.lo
  CC       src/libreadstat_la-readstat_dta_write.lo
  CC       src/libreadstat_la-readstat_error.lo
  CC       src/libreadstat_la-readstat_io_unistd.lo
  CC       src/libreadstat_la-readstat_parser.lo
  CC       src/libreadstat_la-readstat_por.lo
  CC       src/libreadstat_la-readstat_por_parse.lo
  CC       src/libreadstat_la-readstat_rdata.lo
  CC       src/libreadstat_la-readstat_sas.lo
  CC       src/libreadstat_la-readstat_sas_catalog.lo
  CC       src/libreadstat_la-readstat_sas_data.lo
src/readstat_sas_data.c: In function 'sas_parse_subheader_rle':
src/readstat_sas_data.c:505:21: warning: format '%ld' expects argument of type 'long int', but argument 5 has type 'long long int' [-Wformat=]
                     "ReadStat: Row #%d decompressed to %ld bytes (expected %d bytes)\n",
                     ^
  CC       src/libreadstat_la-readstat_sav.lo
  CC       src/libreadstat_la-readstat_sav_parse.lo
  CC       src/libreadstat_la-readstat_sav_read.lo
  CC       src/libreadstat_la-readstat_sav_write.lo
  CC       src/libreadstat_la-readstat_spss.lo
  CC       src/libreadstat_la-readstat_spss_parse.lo
  CC       src/libreadstat_la-readstat_value.lo
  CC       src/libreadstat_la-readstat_variable.lo
  CC       src/libreadstat_la-readstat_writer.lo
  CCLD     libreadstat.la

*** Warning: Trying to link with static lib archive libiconv.a.
*** I have the capability to make that library automatically link in when
*** you link to this library.  But I can only do this if you have a
*** shared version of the library, which you do not appear to have
*** because the file extensions .a of this argument makes me believe
*** that it is just a static archive that I should not use here.
src/.libs/libreadstat_la-readstat_convert.o: In function `readstat_convert':
C:\msys64\home\anthoff\ReadStat/src/readstat_convert.c:24: undefined reference to `libiconv'
src/.libs/libreadstat_la-readstat_dta.o: In function `dta_ctx_init':
C:\msys64\home\anthoff\ReadStat/src/readstat_dta.c:120: undefined reference to `libiconv_open'
C:\msys64\home\anthoff\ReadStat/src/readstat_dta.c:122: undefined reference to `libiconv_open'
src/.libs/libreadstat_la-readstat_dta.o: In function `dta_ctx_free':
C:\msys64\home\anthoff\ReadStat/src/readstat_dta.c:184: undefined reference to `libiconv_close'
src/.libs/libreadstat_la-readstat_sas_catalog.o: In function `sas_catalog_ctx_free':
C:\msys64\home\anthoff\ReadStat/src/readstat_sas_catalog.c:33: undefined reference to `libiconv_close'
src/.libs/libreadstat_la-readstat_sas_catalog.o: In function `readstat_parse_sas7bcat':
C:\msys64\home\anthoff\ReadStat/src/readstat_sas_catalog.c:283: undefined reference to `libiconv_open'
src/.libs/libreadstat_la-readstat_sas_data.o: In function `sas_ctx_free':
C:\msys64\home\anthoff\ReadStat/src/readstat_sas_data.c:135: undefined reference to `libiconv_close'
src/.libs/libreadstat_la-readstat_sas_data.o: In function `readstat_parse_sas7bdat':
C:\msys64\home\anthoff\ReadStat/src/readstat_sas_data.c:1039: undefined reference to `libiconv_open'
src/.libs/libreadstat_la-readstat_sav.o: In function `sav_ctx_free':
C:\msys64\home\anthoff\ReadStat/src/readstat_sav.c:68: undefined reference to `libiconv_close'
src/.libs/libreadstat_la-readstat_sav_parse.o: In function `sav_parse_long_variable_names_record':
C:\msys64\home\anthoff\ReadStat/src/readstat_sav_parse.rl:66: undefined reference to `libiconv'
src/.libs/libreadstat_la-readstat_sav_parse.o: In function `sav_parse_very_long_string_record':
C:\msys64\home\anthoff\ReadStat/src/readstat_sav_parse.rl:173: undefined reference to `libiconv'
src/.libs/libreadstat_la-readstat_sav_read.o: In function `sav_parse_machine_integer_info_record':
C:\msys64\home\anthoff\ReadStat/src/readstat_sav_read.c:855: undefined reference to `libiconv_open'
collect2.exe: error: ld returned 1 exit status
make: *** [Makefile:555: libreadstat.la] Error 1
````

I also tried to paste in the full path to the ``libiconv.a`` file, also no luck.

**Comment by davidanthoff** [comment]
Oh, and for now I'm trying this in my local msys2 environment, easier to first get it to run there than on opensuse.

**Comment by tkelman** [comment]
What's the full link line? `make V=1`

**Comment by davidanthoff** [comment]
````
$ make V=1
/bin/sh ./libtool  --tag=CC   --mode=link gcc -Wall -g -O2 -no-undefined  -o libreadstat.la -rpath /usr/local/lib src/libreadstat_la-CKHashTable.lo src/libreadstat_la-readstat_bits.lo src/libreadstat_la-readstat_convert.lo src/libreadstat_la-readstat_dta.lo src/libreadstat_la-readstat_dta_read.lo src/libreadstat_la-readstat_dta_write.lo src/libreadstat_la-readstat_error.lo src/libreadstat_la-readstat_io_unistd.lo src/libreadstat_la-readstat_parser.lo src/libreadstat_la-readstat_por.lo src/libreadstat_la-readstat_por_parse.lo src/libreadstat_la-readstat_rdata.lo src/libreadstat_la-readstat_sas.lo src/libreadstat_la-readstat_sas_catalog.lo src/libreadstat_la-readstat_sas_data.lo src/libreadstat_la-readstat_sav.lo src/libreadstat_la-readstat_sav_parse.lo src/libreadstat_la-readstat_sav_read.lo src/libreadstat_la-readstat_sav_write.lo src/libreadstat_la-readstat_spss.lo src/libreadstat_la-readstat_spss_parse.lo src/libreadstat_la-readstat_value.lo src/libreadstat_la-readstat_variable.lo src/libreadstat_la-readstat_writer.lo -llzma -lz libiconv.a

*** Warning: Trying to link with static lib archive libiconv.a.
*** I have the capability to make that library automatically link in when
*** you link to this library.  But I can only do this if you have a
*** shared version of the library, which you do not appear to have
*** because the file extensions .a of this argument makes me believe
*** that it is just a static archive that I should not use here.
libtool: link: rm -fr  .libs/libreadstat.dll.a
libtool: link: gcc -shared  src/.libs/libreadstat_la-CKHashTable.o src/.libs/libreadstat_la-readstat_bits.o src/.libs/libreadstat_la-readstat_convert.o src/.libs/libreadstat_la-readstat_dta.o src/.libs/libreadstat_la-readstat_dta_read.o src/.libs/libreadstat_la-readstat_dta_write.o src/.libs/libreadstat_la-readstat_error.o src/.libs/libreadstat_la-readstat_io_unistd.o src/.libs/libreadstat_la-readstat_parser.o src/.libs/libreadstat_la-readstat_por.o src/.libs/libreadstat_la-readstat_por_parse.o src/.libs/libreadstat_la-readstat_rdata.o src/.libs/libreadstat_la-readstat_sas.o src/.libs/libreadstat_la-readstat_sas_catalog.o src/.libs/libreadstat_la-readstat_sas_data.o src/.libs/libreadstat_la-readstat_sav.o src/.libs/libreadstat_la-readstat_sav_parse.o src/.libs/libreadstat_la-readstat_sav_read.o src/.libs/libreadstat_la-readstat_sav_write.o src/.libs/libreadstat_la-readstat_spss.o src/.libs/libreadstat_la-readstat_spss_parse.o src/.libs/libreadstat_la-readstat_value.o src/.libs/libreadstat_la-readstat_variable.o src/.libs/libreadstat_la-readstat_writer.o   -llzma -lz  -g -O2   -o .libs/libreadstat-0.dll -Wl,--enable-auto-image-base -Xlinker --out-implib -Xlinker .libs/libreadstat.dll.a
src/.libs/libreadstat_la-readstat_convert.o: In function `readstat_convert':
C:\msys64\home\anthoff\ReadStat/src/readstat_convert.c:24: undefined reference to `libiconv'
src/.libs/libreadstat_la-readstat_dta.o: In function `dta_ctx_init':
C:\msys64\home\anthoff\ReadStat/src/readstat_dta.c:120: undefined reference to `libiconv_open'
C:\msys64\home\anthoff\ReadStat/src/readstat_dta.c:122: undefined reference to `libiconv_open'
src/.libs/libreadstat_la-readstat_dta.o: In function `dta_ctx_free':
C:\msys64\home\anthoff\ReadStat/src/readstat_dta.c:184: undefined reference to `libiconv_close'
src/.libs/libreadstat_la-readstat_sas_catalog.o: In function `sas_catalog_ctx_free':
C:\msys64\home\anthoff\ReadStat/src/readstat_sas_catalog.c:33: undefined reference to `libiconv_close'
src/.libs/libreadstat_la-readstat_sas_catalog.o: In function `readstat_parse_sas7bcat':
C:\msys64\home\anthoff\ReadStat/src/readstat_sas_catalog.c:283: undefined reference to `libiconv_open'
src/.libs/libreadstat_la-readstat_sas_data.o: In function `sas_ctx_free':
C:\msys64\home\anthoff\ReadStat/src/readstat_sas_data.c:135: undefined reference to `libiconv_close'
src/.libs/libreadstat_la-readstat_sas_data.o: In function `readstat_parse_sas7bdat':
C:\msys64\home\anthoff\ReadStat/src/readstat_sas_data.c:1039: undefined reference to `libiconv_open'
src/.libs/libreadstat_la-readstat_sav.o: In function `sav_ctx_free':
C:\msys64\home\anthoff\ReadStat/src/readstat_sav.c:68: undefined reference to `libiconv_close'
src/.libs/libreadstat_la-readstat_sav_parse.o: In function `sav_parse_long_variable_names_record':
C:\msys64\home\anthoff\ReadStat/src/readstat_sav_parse.rl:66: undefined reference to `libiconv'
src/.libs/libreadstat_la-readstat_sav_parse.o: In function `sav_parse_very_long_string_record':
C:\msys64\home\anthoff\ReadStat/src/readstat_sav_parse.rl:173: undefined reference to `libiconv'
src/.libs/libreadstat_la-readstat_sav_read.o: In function `sav_parse_machine_integer_info_record':
C:\msys64\home\anthoff\ReadStat/src/readstat_sav_read.c:855: undefined reference to `libiconv_open'
collect2.exe: error: ld returned 1 exit status
make: *** [Makefile:555: libreadstat.la] Error 1
````

**Comment by tkelman** [comment]
where is `libiconv.a` located? I think it'll need a full absolute path?

**Comment by davidanthoff** [comment]
I get the same error when I put in the full path:
````
$ make
  CC       src/libreadstat_la-CKHashTable.lo
  CC       src/libreadstat_la-readstat_bits.lo
  CC       src/libreadstat_la-readstat_convert.lo
  CC       src/libreadstat_la-readstat_dta.lo
  CC       src/libreadstat_la-readstat_dta_read.lo
  CC       src/libreadstat_la-readstat_dta_write.lo
  CC       src/libreadstat_la-readstat_error.lo
  CC       src/libreadstat_la-readstat_io_unistd.lo
  CC       src/libreadstat_la-readstat_parser.lo
  CC       src/libreadstat_la-readstat_por.lo
  CC       src/libreadstat_la-readstat_por_parse.lo
  CC       src/libreadstat_la-readstat_rdata.lo
  CC       src/libreadstat_la-readstat_sas.lo
  CC       src/libreadstat_la-readstat_sas_catalog.lo
  CC       src/libreadstat_la-readstat_sas_data.lo
src/readstat_sas_data.c: In function 'sas_parse_subheader_rle':
src/readstat_sas_data.c:505:21: warning: format '%ld' expects argument of type 'long int', but argument 5 has type 'long long int' [-Wformat=]
                     "ReadStat: Row #%d decompressed to %ld bytes (expected %d bytes)\n",
                     ^
  CC       src/libreadstat_la-readstat_sav.lo
  CC       src/libreadstat_la-readstat_sav_parse.lo
  CC       src/libreadstat_la-readstat_sav_read.lo
  CC       src/libreadstat_la-readstat_sav_write.lo
  CC       src/libreadstat_la-readstat_spss.lo
  CC       src/libreadstat_la-readstat_spss_parse.lo
  CC       src/libreadstat_la-readstat_value.lo
  CC       src/libreadstat_la-readstat_variable.lo
  CC       src/libreadstat_la-readstat_writer.lo
  CCLD     libreadstat.la

*** Warning: Trying to link with static lib archive /mingw64/lib/libiconv.a.
*** I have the capability to make that library automatically link in when
*** you link to this library.  But I can only do this if you have a
*** shared version of the library, which you do not appear to have
*** because the file extensions .a of this argument makes me believe
*** that it is just a static archive that I should not use here.
src/.libs/libreadstat_la-readstat_convert.o: In function `readstat_convert':
C:\msys64\home\anthoff\ReadStat/src/readstat_convert.c:24: undefined reference to `libiconv'
src/.libs/libreadstat_la-readstat_dta.o: In function `dta_ctx_init':
C:\msys64\home\anthoff\ReadStat/src/readstat_dta.c:120: undefined reference to `libiconv_open'
C:\msys64\home\anthoff\ReadStat/src/readstat_dta.c:122: undefined reference to `libiconv_open'
src/.libs/libreadstat_la-readstat_dta.o: In function `dta_ctx_free':
C:\msys64\home\anthoff\ReadStat/src/readstat_dta.c:184: undefined reference to `libiconv_close'
src/.libs/libreadstat_la-readstat_sas_catalog.o: In function `sas_catalog_ctx_free':
C:\msys64\home\anthoff\ReadStat/src/readstat_sas_catalog.c:33: undefined reference to `libiconv_close'
src/.libs/libreadstat_la-readstat_sas_catalog.o: In function `readstat_parse_sas7bcat':
C:\msys64\home\anthoff\ReadStat/src/readstat_sas_catalog.c:283: undefined reference to `libiconv_open'
src/.libs/libreadstat_la-readstat_sas_data.o: In function `sas_ctx_free':
C:\msys64\home\anthoff\ReadStat/src/readstat_sas_data.c:135: undefined reference to `libiconv_close'
src/.libs/libreadstat_la-readstat_sas_data.o: In function `readstat_parse_sas7bdat':
C:\msys64\home\anthoff\ReadStat/src/readstat_sas_data.c:1039: undefined reference to `libiconv_open'
src/.libs/libreadstat_la-readstat_sav.o: In function `sav_ctx_free':
C:\msys64\home\anthoff\ReadStat/src/readstat_sav.c:68: undefined reference to `libiconv_close'
src/.libs/libreadstat_la-readstat_sav_parse.o: In function `sav_parse_long_variable_names_record':
C:\msys64\home\anthoff\ReadStat/src/readstat_sav_parse.rl:66: undefined reference to `libiconv'
src/.libs/libreadstat_la-readstat_sav_parse.o: In function `sav_parse_very_long_string_record':
C:\msys64\home\anthoff\ReadStat/src/readstat_sav_parse.rl:173: undefined reference to `libiconv'
src/.libs/libreadstat_la-readstat_sav_read.o: In function `sav_parse_machine_integer_info_record':
C:\msys64\home\anthoff\ReadStat/src/readstat_sav_read.c:855: undefined reference to `libiconv_open'
collect2.exe: error: ld returned 1 exit status
make: *** [Makefile:555: libreadstat.la] Error 1
````
That path should be ok and the file exists there...

**Comment by tkelman** [comment]
Libtool is being difficult
```
*** because the file extensions .a of this argument makes me believe
*** that it is just a static archive that I should not use here.
```
I'm not sure how to make it change its mind about the "should not use" part.

**Comment by davidanthoff** [comment]
I'm tempted to just write a manual makefile, just for the OpenSuse build and just ignore the whole autotools stuff... The makefile was actually *super* easy before the whole autotools stuff was added... And I can get all the static linking etc. going with it as well...

**Comment by evanmiller** [comment]
I'm not sure why you want static linking? The error message earlier in this thread suggests you just need to provide path to libiconv.dll to the linker.

**Comment by davidanthoff** [comment]
@evanmiller: When I build locally (or on appveyor) with msys2 I am using the gnu libiconv implementation right now and everything works using the shared library libiconv.dll. That is not available, though, on the opensuse build server. Instead, they have win_iconv, an alternative implementation. And unfortunately then opensuse people are [insisting](https://build.opensuse.org/request/show/402827) that one links win_iconv statically.

So, I guess our options are this:
1) Try to convince the opensuse people to not delete the libiconv.dll.a in their build. They actually produce it and then manually delete it... In that case we can use the iconv.dll from the win_iconv project.
2) Host our own fork of win_iconv that doesn't delete the libiconv.dll. a file. Annoying because we now have to maintain another repo.
3) Get the build script to statically link win_iconv only on the opensuse build server. That would be the best solution, but I don't know how to achieve it. Maybe [this](https://lists.opensuse.org/opensuse-mingw/2011-04/msg00020.html) can help...
4) Ditch the whole opensuse build service idea, just build libreadstat.dll on appveyor with msys2, and statically link libiconv and zlib into it. At that point libreadstat.dll only depends on Windows dlls, no other mingw compiled dlls. @tkelman we shouldn't have problems with different ABIs in that case, right? The drawback is that the libreadstat.dll then is about 1.5 MB... But the build script is simple, it is essentially a one-liner gcc call...

Any other ideas?

**Comment by tkelman** [comment]
Depends what the ccall's need to accomplish exactly. If it's pure C then msys2's gcc should work, they are built with different exception handling and threading models than what we use on opensuse for winrpm and cross-compiled from Cygwin for Julia itself.

The libtool cache variable thing is something I've seen in a decent number of packages on the opensuse build service, good to have a reference on what that might be for.

**Comment by davidanthoff** [comment]
@jar1karp, I'm wondering whether you might be able to help here, you added the whole autotools build stuff, right? I'm trying to build ReadStat on the opensuse build server for windows so that we can use it easily from julia. On that service, we can't use GNU libiconv, but instead must use win_iconv. And we must link win_iconv statically into the libreadstat-0.dll because the mingw64 build on opensuse prescribes that.

The problem is that I really don't understand autotools at all, and I'm just lost on how to achieve this. Any help would be greatly appreciated!

The opensuse project is here: https://build.opensuse.org/package/show/home:davidanthoff/mingw64-readstat

**Comment by jar1karp** [comment]
@davidanthoff Ok, I've branched the OBS project and it seemed to build automatically. I didn't get any errors and after enabling publishing the built packages became available at http://download.opensuse.org/repositories/home:/jar1karp:/branches:/home:/davidanthoff/openSUSE_Leap_42.1/noarch/

The build seems to be working ok (no build errors, just some minor rpmlint warnings).

**Comment by tkelman** [comment]
Did you change anything other than `echo "lt_cv_deplibs_check_method='pass_all'" >>%{_mingw64_cache}`, or was that already there? (OBS is down for maintenance at the moment, otherwise the repo history would probably answer this)

**Comment by jar1karp** [comment]
I didn't make any changes to the spec.

**Comment by jar1karp** [comment]
Currently mingw64 build produces libreadstat-0.dll which DataRead.jl is apparently unable to call as it expects to find libreadstat.dll. The above PR should fix that problem.

I did some additional testing, my changes are at https://build.opensuse.org/package/show/home:jar1karp:branches:home:davidanthoff/mingw64-readstat

Seems that `echo "lt_cv_deplibs_check_method='pass_all'" >>%{_mingw64_cache}` is not needed, the build will work fine as long as win_iconv and zlib1 are in BuildRequires. WinRPM will also install both.

After adding the mingw64 bin directory to DLL path I was able to get Stata datafiles loaded in Julia.

**Comment by tkelman** [comment]
It would be better to fix DataRead.jl to handle an alias rather than adding `-avoid-version` here.

**Comment by davidanthoff** [comment]
@jar1karp Thanks for your help, much appreciated!

The DLL name is not an issue, in fact I had adjusted the build.jl in DataRead.jl already to handle the versioned file name, see [here](https://github.com/WizardMac/DataRead.jl/pull/13).

The problem is that opensuse doesn't distribute a dll of win_iconv, they want people to statically link win_iconv instead. The build I've created on build.opensuse.org does compile and produces a libreadstat.dll, but that dll has a dependency on a iconv.dll, and we won't get that from WinRPM because it is not being distributed via opensuse. I *assume* that things worked on your machine because you added the mingw64 bin directory to your path and had a iconv.dll (or libiconv.dll) in there. But that won't work for us because we should get this to work for people that don't have mingw64 or anything like that installed.

If we are sticking with the opensuse stuff, we need to compile things on their infrastructure such that win_iconv is statically linked into libreadstat.dll. It is not clear to me how to do this. I.e. what we want is a libreadstat.dll that has the win_iconv stuff linked into libreadstat.dll, and then has a dependency on zlib1.dll. Maybe an easier option would be to just link everything statically into libreadstat.dll, i.e. both win_iconv and zlib1, so that libreadstat.dll has no dependency on any other dll?

**Comment by davidanthoff** [comment]
See also [here](https://build.opensuse.org/request/show/402827).

**Comment by jar1karp** [comment]
Ok, I guess the PR #66 could be reverted then.

Linking win_iconv statically seems to work, [here's](https://github.com/jar1karp/ReadStat/commit/6a9628239a62a7c6555cca5879072924dbcc0e3e) a commit that does this by passing static linking flag to linker.

Looking at the OBS build log then built package has only a dependency to zlib.

Basically the same could be done for zlib, but [the spec](https://build.opensuse.org/package/view_file/windows:mingw:win64/mingw64-zlib/mingw64-zlib.spec?expand=1) explicitly removes the static lib, so that would need to be changed.

**Comment by davidanthoff** [comment]
@jar1karp We are getting very close, when I use your commit, things look good: I get a libreadstat.dll that depends on zlib.dll only. That is exactly what we want because zlib.dll is actually distributed via an RPM, so we wouldn't want to link that statically.

I think it would make most sense though to link iconv.dll statically *only* in the opensuse build, but not in the msys2 or cygwin builds. I think for the latter two we are using GNU iconv which is large, so we wouldn't want to link that statically. Only on opensuse do we end up using win_iconv, which is small and thus makes sense to link statically.

I tried to incorporate the flags from your commit into the [spec file](https://build.opensuse.org/package/view_file/home:davidanthoff/mingw64-readstat/mingw64-readstat.spec?expand=1) instead of modifying the actual build files here in the repo, but I'm doing something wrong, so far this produces rpms that don't link icon statically. Right now I'm trying to pass these flags to the make step, but I also tried to pass them to the configure step. Or do I have to pass them to both? I might also be using the wrong syntax in the spec file for this. @tkelman any hint?

**Comment by tkelman** [comment]
Probably easiest to just carry local patches to the autotools files rather than working out the right set of flags to make it happy.

**Comment by jar1karp** [comment]
Patch approach is probably simplest solution for now. Another solution would be to add an argument for configure using AC_ARG_ENABLE that enables static linking of iconv.

**Comment by davidanthoff** [comment]
I think there is one more option: the ``Host`` value seems different on opensuse from msys2, so I think we can just add another case in the configure.ac file. On my local msys2 I get `` x86_64-pc-mingw64`` as the host value, in the opensuse log it seems to be ``x86_64-w64-mingw32``. That should be enough to set the different flags, right?

**Comment by tkelman** [comment]
pc-mingw64 is odd and nonstandard, I've never seen that before. msys2 must have changed that recently, adding to my list of grievances about that project.

**Comment by davidanthoff** [comment]
#68 should help.

**Comment by davidanthoff** [comment]
I'll keep the discussion here. We should also get a 32 bit thing up on opensuse. Do I create another project for that, called mingw32-readstat? Or something else?

Also, I don't understand versioning with this RPM stuff. Is there always just one version hosted on RPM?

**Comment by jar1karp** [comment]
[Here's a commit](https://github.com/jar1karp/ReadStat/commit/8d20ba57f84153d6d0058b032ec359370e3364fa) that adds the optional `--enable-static-iconv` flag which can be passed to configure in opensuse spec.

**Comment by tkelman** [comment]
yes to both your questions. on 32 bit, just replace all 64's with 32 in the spec, and replace all x86_64 with i686. the versioning is unfortunately a "latest release" deal, you'd have to do the annoying homebrew thing of separatelyn named  old version packages to keep them around I think. there's a better approach to package manager versioning out there but I haven't quite found it yet.

**Comment by davidanthoff** [comment]
Alright, another question: right now I build on openSUSE_13.2 and openSUSE_Leap_42.1. Do I need both, or can I remove one of them? My sense is that openSUSE_13.2 should be enough, right?

**Comment by tkelman** [comment]
Right now WinRPM downloads only from 13.2, but that's in danger of getting replaced whenever opensuse 42.2 or whatever comes out so it may all stop working. We could actually switch WinRPM to use 42.1 now and I don't think anyone would notice, since the whole cross compiled stack gets built with the same versions across the different linux distro versions numbers.

**Comment by davidanthoff** [comment]
@jar1karp Cool, that is perfect. I just tried it and it works on opensuse. I opened #69 to merge this into master, ok?

**Comment by jar1karp** [comment]
Ok, cool.

**Comment by davidanthoff** [comment]
@tkelman Most of the builds have been "blocked" for more than 22 hours now. Does that sometimes happen, i.e. the whole opensuse infrastructure gets blocked for that long? Or do I need to trigger a new build somehow?

**Comment by tkelman** [comment]
see https://build.opensuse.org/project/monitor/windows:mingw:win32, when something at the bottom of the stack gets changed then everything else rebuilds, yeah.

**Comment by davidanthoff** [comment]
I think the complicated stuff is all done now, just a bunch of smallish things left do.

- [ ] @evanmiller any chance you could tag a v0.1.0 now? It would be good if we could sort out how versioning will work here and make sure the scripts all work with a tarball that is downloaded from a github tag etc. A v0.1.0 would essentially just be a practice release for us that allows me to make sure everything works as expected once version info is there. @jar1karp this might also require some more changes in the autotools stuff so that the version number is somehow used there as well? Not clear to me how these things work.
- [ ] I would prefer if the binaries were not hosted under my account on opensuse. It would probably make more sense if there was either a ``home::WizardMac`` project that hosted this stuff if @evanmiller wants to maintain this going forward, or maybe @tkelman could set up a project ``julialang`` that hosts various binary packages that are needed by julia packages?

The model for updating the build on opensuse would now be this: download a tarball from github from a tagged version, upload it to https://build.opensuse.org/package/show/home:davidanthoff/mingw64-readstat and https://build.opensuse.org/package/show/home:davidanthoff/mingw32-readstat, edit the spec files in each location to match the version, and then everything should compile. That is still a bunch of manual steps, but I think we'll just leave it for now like that?

**Comment by tkelman** [comment]
updating just the tarball through a web interface is probably easier than having to compile and upload windows binaries yourself, especially if you primarily use mac where the mingw cross compilers are a bit complicated to install.

**Comment by jar1karp** [comment]
Re: versioning, I have [a branch](https://github.com/jar1karp/ReadStat/commits/autotools-improvements) with some initial commits to do this.

**Comment by davidanthoff** [comment]
Alright, everything seems to work!! Hurray :)

This just leaves the question what account on opensuse should host the binaries. @evanmiller do you intent to maintain the windows binaries on opensuse for the julia package? If so, it would probably make most sense if they were hosted under an opensuse account in your name. If not, @tkelman any chance we could set up some julialang like opensuse repo for these kind of things?

**Comment by tkelman** [comment]
I don't know how to set up an organization on the opensuse build service, but it's probably possible. we could also try submitting to the official mingw project at some point, then no messing with WinRPM.sources would be needed

**Comment by davidanthoff** [comment]
Yeah, I kind of thought if there was an official julialang account on opensuse, that could just be added to WinRPM.sources by default... Maybe just create a personal account named julialang that julia computing maintains? I don't really understand the concepts that are around on opensuse in any case.

**Comment by evanmiller** [comment]
@davidanthoff I don't really have any interest in maintaining binaries on opensuse.

**Comment by davidanthoff** [comment]
@tkelman Any thoughts about an official julialang project on build.opensuse.org that could host binaries like this one here? Might also be useful for e.g. Rmath.jl stuff? When I click on the "New Project" button on build.opensuse.org, it says "All users have a homes project by default. Yours is home:davidanthoff, and you can also create sub projects inside. New projects outside of your home project can get requested via email." with a link to the email, so that might be an option.

**Comment by tkelman** [comment]
I don't have time to worry about that right now, and wasn't planning on using their build service for rmath

**Comment by davidanthoff** [comment]
Alright, I'll host it for the time being. But would be great if we could revisit this at some point.

Thanks everyone for all the help, there is no way that I could have pulled this off by myself!
